_posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md

```diff
@@ -0,0 +1,93 @@
+}
+```
+
+In the test above, if the call to `decompose` ever causes any overflow, underflow, division by zero, etc. then Kani will report an error and the exact value and bits inputted that caused the problem. This feedback is great, because one can just plugin those values into another test and debug it precisely. One will note that this test does not actually check the output value. That's for two reasons. First, testing values for all possible combinations would require re-implementing the same code that is in the `decompose` function making the test a tautology. Second, the purpose of the Hifitime Kani tests is to ensure that the aren't any unsound operations. Moreover, Hifitime has plenty of values that are explicitly tested for in the rest the tests.
+
+## Conclusion
+
+Kani has helped fix at least eight different categories of bugs in a single pull request: <https://github.com/nyx-space/hifitime/pull/192>. Most of these were bugs near the boundaries of a Duration definition: around the zero, maximum, and minimum durations. But many of the bugs were on important and common operations: partial equality, negation, addition, and subtraction operations. These bugs weren't due to lax testing: there are over 74 integration tests with plenty of checks within each.
+
+One of the great features of Kani is that it performs what is known as symbolic execution of programs, where inputs are modelled as symbolic variables covering whole ranges of values at once. All program behaviors possible under these inputs are analyzed for defects like arithmetic overflows or underflows, signed conversion overflow or underflow, etc.. If a defect is possible for some values of the inputs, Kani will generate a counter example trace with concrete values triggering the defect.
+
+Thanks to how Kani analyzes a program, tests can either be bounded or not. A bounded test includes an assertion: execute a set of instructions and then check something. This is a typical test case.
+
+Kani can also have unbounded tests where there is no explicit condition to check. Instead, only the successive operations of a function call are executed, and each are tested by Kani for failure cases by analyzing the inputs and finding cases where inputs will lead to runtime errors like overflows. This approach is how most of the bugs in hifitime have been found.
+
+Unbounded tests effectively ensure that sanity of the operations in a given function call. Bounded tests provide the same while also checking equalities. If either of these tests fail, Kani can provide a test failure report outlining the sequence of operations, and the binary representation of each intermediate operation, to help the developer gain an understanding of why their implementation is incorrect.
+
+The overhead to implement tests in Kani is very low, and the benefits are immense. Basically, write a Kani verification like a unit test, run the model verifier, and you've formally verified this part of the code. Amazing!
```

Comment:
Could be useful to mention the number of Kani tests that are integrated in `hifitime`.